### DISCUSSION TITLE
Relationship between domains, organizations, services and ssl certs and how it impacts deployment

### DISCUSSION THREAD

**Discussion by seanrmurphy** [magma/magma#5299] [General]
I have been working through a test setup and I've come to the realization that there are linkages between domain names, organizations, services and ssl certs which are not very explicit and I think it would be good to clarify them. My test setup is neither AWS nor bare metal so the I did not use the full automation suite; also I want to understand a bit more clearly what is going on - I'm working in Openstack VMs right now.

My understanding is as follows:
- a basic premise of Magma is that a single Magma deployment can support multiple distinct Organizations which can be completely independent networks - potentially comprising different technologies, serving different customers and having different objectives
- every Magma deployment has a `master` Organization - this is essentially the system administrator and can be used to add new Organizations
- Magma uses domain names to determine which Organization the user is trying to interact with
  -  for example, I think there is an email matching filter here which checks if the site's domain and the email address match
- By design, Magma should expose some services/endpoints specific to different Organizations
  - I guess the intention is to support, for example, `fluentd.<magma-organization>.<magma-hosting-domain>.<tld>` where `<magma-hosting-domain>.<tld>` is constant across all Organizations defined on the magma instance and `<magma-organization>.<magma-hosting-domain>.<tld>` is the entry poiint for one specific Organization

I encountered issues specifically on [this line](https://github.com/magma/magma/blob/master/orc8r/cloud/docker/nginx/templates/nginx.conf.j2#L120) in the `nginx.conf` file, during the bootstrapping process. My problem was that I did not specify any `controller_hostname` and hence this regex matching did not work for me - I'm still not certain how to fix this, although I guess I can work it out for the simple case; however, it gave rise to the question of how this works in the context of different Organizations with different domain names.

So, I have two specific questions:
- what is the minimum set of certificates which is required to set up a test deployment with a `master` and a single Organization
  - do any specific values need to be provided with the `helm install`  stipulating eg that the `master` should be accessible on `master.mydomain.com` and the Organization should be accessed on `*.test-organization.mydomain.com`
  - does the `master` domain need a wildcarded SSL cert?
- for the more general case where there are multiple domains, does this work with the `controller_hostname` specified in the line above or is it necessary to make some more significant changes to the nginx configuration in this case eg adding specific blocks for each Organization?

A further complication in my case is that I'm 'only' using minikube with port-forwarding; perhaps such issues can be more easily managed using a proper ingress on a k8s cluster. 

In any case, I think it would be helpful generally to clarify the relationships assumed between these entities within Magma - any pointers appreciated.

**Comment by karthiksubraveti** [comment]
@seanrmurphy  This topic is briefly discussed here - https://docs.magmacore.org/docs/next/nms/setup. 
Your understanding is correct that we use subdomain name to identify the organization. Organization is still a not a first class construct in Magma. Only NMS is organization aware and does the job of restricting organization to specific networks. As far as the certs are concerned, NMS needs only an admin cert to access Orc8r.  Orc8r is only aware of networks at this point. It doesn't have any concept of multitenancy.

In case of bootstrapping, for e.g. we have the following control_proxy yml in our staging setup.
The subdomain refers more to the various services running on Orc8r. 
```
magma@magma:~$ cat /var/opt/magma/configs/control_proxy.yml

cloud_address: controller-staging.magma.etagecom.io
cloud_port: 443

bootstrap_address: bootstrapper-controller-staging.magma.etagecom.io
bootstrap_port: 443

fluentd_address: fluentd-staging.magma.etagecom.io
fluentd_port: 24224

rootca_cert: /var/opt/magma/certs/rootCA.pem
```
We need following certs in single organization or multi organization deployments
- root certificates
- Controller Certificate (controller.key, controller.crt): 
- Admin Certificate (admin_operator.key.pem, admin_operator.pem) (for API access and connecting from NMS)
- Fluentd Certificate (fluentd.key, fluentd.pem)
- Certifier Certificate (certifier.key, certifier.pem)
- Bootstrapper key

Let me know if you have any further questions here. Thanks

**Reply by hcgatewood** [reply]
For more context on Orc8r e2e security, see https://github.com/magma/magma/pull/5349

**Reply by seanrmurphy** [reply]
Thanks for the clarifications @karthiksubraveti - just to make sure I understand....

As noted above, I thought the idea was to have endpoints named `<service>.<organization-name>.<magma-hosting-domain>.<tld>` and each Organization would have its own (named) services - this would required a wildcard SSL cert for each Organization,obviously (and of course management of that). I think this was a misunderstanding.

From your explanation it seems that the idea is to have the same DNS namespace for both services and Organizations such that each Organization will have an NMS at `<organization-name>.<magma-hosting-domain>.<tld>` and all of the services will be shared across all Organizations and named `<service>.<magma-hosting-domain>.<tld>`. (I guess this means that it will not be possible to have an Organization with the same name as a service).

Of course I understand that the ideas could be evolving etc and things may change as Magma evolves, but is that an accurate characterization of the current state?
